File: android/src/main/java/im/zego/zego_express_engine/ZegoCustomVideoRenderManager.java
package im.zego.zego_express_engine;

import java.nio.ByteBuffer;

import im.zego.zegoexpress.callback.IZegoCustomVideoRenderHandler;
import im.zego.zegoexpress.constants.ZegoPublishChannel;
import im.zego.zegoexpress.constants.ZegoVideoFlipMode;
import im.zego.zegoexpress.entity.ZegoVideoEncodedFrameParam;
import im.zego.zegoexpress.entity.ZegoVideoFrameParam;

public class ZegoCustomVideoRenderManager extends IZegoCustomVideoRenderHandler {

    private volatile static ZegoCustomVideoRenderManager singleton;

    private static IZegoFlutterCustomVideoRenderHandler mHander;

    public ZegoCustomVideoRenderManager() {
    }

    /**
     * Get the custom video capture manager instance
     */
    public static synchronized ZegoCustomVideoRenderManager getInstance() {
        if (singleton == null) {
            singleton = new ZegoCustomVideoRenderManager();
        }
        return singleton;
    }

    /**
     * Sets up the event callback handler for custom video rendering.
     *
     * Available since: 1.9.0
     * Description: Custom video rendering, where the developer is responsible for rendering the video data onto the UI components. With the change callback set, a callback that throws video data to the developer will be triggered when the developer calls startPreview, startPublishingStream, and startPlayingStream.
     * Use case: Use beauty features or apps that use a cross-platform UI framework (for example, Qt requires a complex hierarchical UI to achieve high-experience interaction) or game engine (e.g. Unity, Unreal Engine, Cocos)
     * Related callbacks: [onCapturedVideoFrameRawData] local preview video frame data callback, [onRemoteVideoFrameRawData] remote playing stream video frame data callback.
     * Note: This function is only available in ZegoExpressVideo SDK!
     *
     * @param handler Custom video render handler
     */
    public void setCustomVideoRenderHandler(IZegoFlutterCustomVideoRenderHandler handler) {
        mHander = handler;
    }

    @Override
    public void onCapturedVideoFrameRawData(ByteBuffer[] data, int[] dataLength, ZegoVideoFrameParam param, ZegoVideoFlipMode flipMode, ZegoPublishChannel channel) {
        if (mHander != null) {
            ZGFlutterVideoFrameParam videoFrameParam = new ZGFlutterVideoFrameParam();
            videoFrameParam.format = ZGFlutterVideoFrameFormat.getVideoFrameFormat(param.format.value());
            videoFrameParam.height = param.height;
            videoFrameParam.width = param.width;
            videoFrameParam.rotation = param.rotation;
            for (int i = 0; i < 4; i++) {
                videoFrameParam.strides[i] = param.strides[i];
            }

            mHander.onCapturedVideoFrameRawData(data, dataLength, videoFrameParam, ZGFlutterVideoFlipMode.getZegoVideoFlipMode(flipMode.value()), ZGFlutterPublishChannel.getZegoPublishChannel(channel.value()));
        }
    }

    @Override
    public void onRemoteVideoFrameRawData(ByteBuffer[] data, int[] dataLength, ZegoVideoFrameParam param, String streamID) {
        if (mHander != null) {
            ZGFlutterVideoFrameParam videoFrameParam = new ZGFlutterVideoFrameParam();
            videoFrameParam.format = ZGFlutterVideoFrameFormat.getVideoFrameFormat(param.format.value());
            videoFrameParam.height = param.height;
            videoFrameParam.width = param.width;
            videoFrameParam.rotation = param.rotation;
            for (int i = 0; i < 4; i++) {
                videoFrameParam.strides[i] = param.strides[i];
            }

            mHander.onRemoteVideoFrameRawData(data, dataLength, videoFrameParam, streamID);
        }
    }

    @Override
    public void onRemoteVideoFrameEncodedData(ByteBuffer data, int dataLength, ZegoVideoEncodedFrameParam param, long referenceTimeMillisecond, String streamID) {
        if (mHander != null) {
            ZGFlutterVideoEncodedFrameParam videoEncodedFrameParam = new ZGFlutterVideoEncodedFrameParam();
            videoEncodedFrameParam.isKeyFrame = param.isKeyFrame;
            videoEncodedFrameParam.SEIData = param.SEIData;
            videoEncodedFrameParam.SEIDataLength = param.SEIDataLength;
            videoEncodedFrameParam.format = ZGFlutterVideoEncodedFrameFormat.getZegoVideoEncodedFrameFormat(param.format.value());
            videoEncodedFrameParam.height = param.height;
            videoEncodedFrameParam.width = param.width;
            videoEncodedFrameParam.rotation = param.rotation;
            mHander.onRemoteVideoFrameEncodedData(data,dataLength, videoEncodedFrameParam, referenceTimeMillisecond, streamID);
        }
    }
}

File: windows/ZegoCustomVideoRenderManager.cpp
#include "include/zego_express_engine/ZegoCustomVideoRenderManager.h"

#include <ZegoExpressSDK.h>

static std::shared_ptr<ZegoCustomVideoRenderManager> instance_ = nullptr;
static std::once_flag singletonFlag;

class ZegoCustomVideoRenderHandler : public ZEGO::EXPRESS::IZegoCustomVideoRenderHandler {
public:
    void setCustomVideoRenderHandler(std::shared_ptr<IZegoFlutterCustomVideoRenderHandler> handler) {
        handler_ = handler;
    }
private:
    std::shared_ptr<IZegoFlutterCustomVideoRenderHandler> handler_ = nullptr;

    void onCapturedVideoFrameRawData(unsigned char ** data,
                                             unsigned int * dataLength,
                                             ZEGO::EXPRESS::ZegoVideoFrameParam param,
                                             ZEGO::EXPRESS::ZegoVideoFlipMode flipMode,
                                             ZEGO::EXPRESS::ZegoPublishChannel channel) override {
        if (handler_) {
            ZGFlutterVideoFrameParam frameParam;
            frameParam.width = param.width;
            frameParam.height = param.height;
            frameParam.rotation = param.rotation;
            frameParam.format = (ZGFlutterVideoFrameFormat)param.format;
            for (size_t i = 0; i < 4; i++)
            {
                frameParam.strides[i] = param.strides[i];
            }
            handler_->onCapturedVideoFrameRawData(data, dataLength, frameParam, (ZGFlutterVideoFlipMode)flipMode, (ZGFlutterPublishChannel)channel);
        }
    }

    void onRemoteVideoFrameRawData(unsigned char ** data, unsigned int * dataLength,
                                           ZEGO::EXPRESS::ZegoVideoFrameParam param,
                                           const std::string & streamID) override {
        if (handler_) {
            ZGFlutterVideoFrameParam frameParam;
            frameParam.width = param.width;
            frameParam.height = param.height;
            frameParam.rotation = param.rotation;
            frameParam.format = (ZGFlutterVideoFrameFormat)param.format;
            for (size_t i = 0; i < 4; i++)
            {
                frameParam.strides[i] = param.strides[i];
            }
            handler_->onRemoteVideoFrameRawData(data, dataLength, frameParam, streamID);
        }
    }

    void onRemoteVideoFrameEncodedData(const unsigned char * data,
                                               unsigned int dataLength,
                                               ZEGO::EXPRESS::ZegoVideoEncodedFrameParam param,
                                               unsigned long long referenceTimeMillisecond,
                                               const std::string & streamID)override {
        if (handler_) {
            ZGFlutterVideoEncodedFrameParam frameParam;
            frameParam.width = param.width;
            frameParam.height = param.height;
            frameParam.rotation = param.rotation;
            frameParam.format = (ZGFlutterVideoEncodedFrameFormat)param.format;
            frameParam.isKeyFrame = param.isKeyFrame;
            frameParam.SEIData = param.SEIData;
            frameParam.SEIDataLength = param.SEIDataLength;
            handler_->onRemoteVideoFrameEncodedData(data, dataLength, frameParam, referenceTimeMillisecond, streamID);
        }
    }
};

std::shared_ptr<ZegoCustomVideoRenderManager> ZegoCustomVideoRenderManager::getInstance() {
    std::call_once(singletonFlag, [&] {
        instance_ = std::make_shared<ZegoCustomVideoRenderManager>();
    });
    return instance_;
}

std::shared_ptr<ZEGO::EXPRESS::IZegoCustomVideoRenderHandler> ZegoCustomVideoRenderManager::getHandler() {
    return zegoHandler_;
}

void ZegoCustomVideoRenderManager::setCustomVideoRenderHandler(std::shared_ptr<IZegoFlutterCustomVideoRenderHandler> handler) {
    handler_ = handler;
    if (handler_ && !zegoHandler_) {
        zegoHandler_ = std::make_shared<ZegoCustomVideoRenderHandler>();
    }
    dynamic_cast<ZegoCustomVideoRenderHandler*>(zegoHandler_.get())->setCustomVideoRenderHandler(handler_);
}